subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -332,30 +410,121 @@ func (h *HivedAlgorithm) initFreeCellList() {
 func (h *HivedAlgorithm) initReservations() {
 	for vc, vcReservation := range h.reservedCells {
 		for rid, physical := range vcReservation {
+			h.vcRemainingQuotas[vc][physical.GetChain()][physical.GetLevel()]--
+			h.totalRemainingQuotas[physical.GetChain()][physical.GetLevel()]--
 			h.removeCellFromFreeList(physical)
 			virtualList := h.vcSchedulers[vc].getReservedCellList()[rid]
 			virtual := virtualList[CellLevel(len(virtualList))][0].(*VirtualCell)
-			virtual.SetPhysicalCell(physical)
-			physical.SetVirtualCell(virtual)
-			klog.Infof("Cells bound: %v and %v (reservation)", virtual.GetName(), physical.GetName())
+			bindCell(physical, virtual)
+		}
+	}
+}
+
+// setBad marks a physical cell as bad and recursively for its parent, guaranteeing that
+// a cell is bad if all of its children are bad. The virtual cell that this physical cell
+// is bound to will also be marked as bad. If the cell has not been bound to a virtual cell,
+// but the scheduler finds that a VC will be allocated such a bad cell,
+// one of the cell in that VC will be marked as bad.
+func (h *HivedAlgorithm) setBad(c *PhysicalCell) {
+	c.GetStatus().State = api.BadState
+	if c.virtualCell != nil {
+		c.virtualCell.GetStatus().State = api.BadState
+		c.virtualCell.GetStatus().PhysicalCell.State = api.BadState
+		c.GetStatus().VirtualCell.State = api.BadState
+	}
+	chain, level := c.GetChain(), c.GetLevel()
+	h.numBadCells[chain][level]++
+	if h.totalRemainingQuotas[chain][level] > h.numLeftCells[chain][level]-h.numBadCells[chain][level] {
+		klog.Warningf("Cell type %v (chain %v level %v) now has fewer healthy cells than "+
+			"the total remaining quota of the VCs (total quota %v, healthy %v, bad %v). "+
```

Comment:
>quota of the VCs (total quota  [](start = 24, length = 30)

Rename quota to be a more concert concept. Such as total/free cells. Also for other places, check word "quota".